libcamera_app.hpp

```diff
@@ -744,20 +715,50 @@ class LibcameraApp
 			}
 			preview_frames_displayed_++;
 			preview_->Show(fd, size, w, h, stride);
-			std::string s = getInfoText(buffer, item.completed_request.metadata);
-			preview_->SetInfoText(s);
+			if (!options.info_text.empty())
+			{
+				std::string s = getInfoText(frame_info);
+				preview_->SetInfoText(s);
+			}
 		}
 	}
-	std::string getInfoText(libcamera::FrameBuffer *buffer, libcamera::ControlList &controls)
+	std::string getInfoText(const FrameInfo &frame_info)
 	{
-		const libcamera::FrameMetadata &metadata = buffer->metadata();
-		float exposure, analogue_gain, digital_gain, colour_gains[2];
-		char s[128];
-
-		GetFrameInfo(controls, &exposure, &analogue_gain, &digital_gain, colour_gains);
-		snprintf(s, sizeof(s), "Frame #%u: Exp %.2fms AnaGain %.2fx DigGain %.2fx RedGain %.2fx BlueGain %.2fx",
-				 metadata.sequence, exposure/1000.0, analogue_gain, digital_gain, colour_gains[0], colour_gains[1]);
-		return std::string(s);
+		// Info text tokens.
+		static const std::string tokens[] = {"%frame", "%fps", "%exp", "%ag", "%dg", "%rg", "%bg", "%focus", "%aelock"};
+		std::string info_string(options.info_text);
+
+		for (auto const t: tokens)
+		{
+			std::size_t pos = info_string.find(t);
+			if (pos != std::string::npos)
+			{
+				std::stringstream value;
+				value << std::fixed << std::setprecision(2);
+
+				if (t == "%frame")
+					value << frame_info.sequence;
+				else if (t == "%fps")
+					value << frame_info.fps;
+				else if (t == "%exp")
+					value << frame_info.exposure_time;
+				else if (t == "%ag")
+					value << frame_info.analogue_gain;
+				else if (t == "%dg")
+					value << frame_info.digital_gain;
+				else if (t == "%rg")
+					value << frame_info.colour_gains[0];
+				else if (t == "%bg")
+					value << frame_info.colour_gains[1];
+				else if (t == "%focus")
+					value << frame_info.focus;
+				else if (t == "%aelock")
+					value << frame_info.aelock;
+
+				info_string.replace(pos, t.length(), value.str());
+			}
+		}
+		return info_string;
 	}
```

Comment:
Looking at this 2nd commit, maybe we could make this a member of the FrameInfo? The LibcameraApp class already has too much in it...!

Reply:
I can add a `to_string()` method that takes in the token string and returns out the parsed string.

Reply:
Except I think method names are CamelCase (or camelCase if private...). So much for being consistent!!